src/auth/src/credentials/service_account.rs

```diff
@@ -32,33 +94,117 @@ use time::OffsetDateTime;
 const DEFAULT_SCOPES: &str = "https://www.googleapis.com/auth/cloud-platform";
 
 pub(crate) fn creds_from(js: serde_json::Value) -> Result<Credential> {
-    let service_account_info =
-        serde_json::from_value::<ServiceAccountInfo>(js).map_err(CredentialError::non_retryable)?;
-    let token_provider = ServiceAccountTokenProvider {
-        service_account_info,
-    };
-    let token_provider = TokenCache::new(token_provider);
-
-    Ok(Credential {
-        inner: Arc::new(ServiceAccountCredential { token_provider }),
-    })
+    let service_account_key =
+        serde_json::from_value::<ServiceAccountKey>(js).map_err(CredentialError::non_retryable)?;
+    Ok(Builder::default()
+        .service_account_key(service_account_key)
+        .build())
+}
+
+#[derive(Default)]
+pub struct Builder {
+    service_account_key: ServiceAccountKey,
+    aud: Option<String>,
+    scopes: Option<String>,
+    quota_project_id: Option<String>,
+}
+
+impl Builder {
+    /// Sets the [service account key] in [aip/4112] format for this credential.
+    ///
+    /// [aip/4112]: https://google.aip.dev/auth/4112
+    /// [service account key]: https://cloud.google.com/iam/docs/keys-create-delete#creating
+    pub fn service_account_key(mut self, service_account_key: ServiceAccountKey) -> Self {
+        self.service_account_key = service_account_key;
+        self
+    }
+
+    /// Sets the audience for this credential.
+    ///
+    /// `aud` is a [JWT] claim specifying intended recipient(s) of the token,
+    /// that is, a service(s) or resource(s).
+    /// This cannot be used at the same time as the scopes claim.
```

Comment:
This suggests we should change the implementation to be something like:

```
enum ServiceAccountRestrictions {
  Audience(String),
  Scopes(Vec<String>),
}

pub struct Builder {
  service_account_key: ServiceAccountKey,
  restrictions: Option<ServiceAccountRestrictions>,
  quota_project_id: Option<String>,
}
```

And then you can change the implementation to:

```
impl Builder {
    pub fn aud<V: Into<String>>(mut self, v: V) -> Self {
        self.restrictions = ServiceAccountRestrictions::Audience(v.into());
        self
    }
}
```

The comments should mention that setting `aud()` overrides the value set in `scopes()`, but now it is impossible to set both.